design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,260 @@
+
+The proposal takes the top-down approach, starting with the high level
+flow that a common feature will traverse through.
+Continuing with actions that need to be taken and timeline suggestions.
+
+Both feature graduation and discontinuation flows are covered.
+Including the implications on users.
+
+Depending on individual topics, follow-up proposals may extend the basic
+points raised in this proposal.
+
+### Feature Stages
+A feature is expected to pass through the following stages:
+1. Enhancement proposal.
+2. Implementation.
+3. Release as Alpha (experimental).
+4. Release as Beta (pre-release for evaluation).
+5. Release as GA (graduation).
+6. Discontinuation.
+
+Starting from the Alpha release, it can be discontinued with restrictions that
+depend on the release stage (Alpha, Beta, GA).
+
+Deprecation of features is widely discussed later in this proposal.
+TODO: Add a ref to the section here.
+
+#### Enhancement proposal
+As the first step for introducing a new feature, a formal proposal is
+expected to be shared for public review.
+
+This is the first opportunity to evaluate a new feature.
+The proposal needs to include motivation, goals, implementation details
+and phases.
+
+#### Implementation
+The development work on the feature is expected to include coding,
+testing, integration and documentation.
+
+While most of the effort is expected to occur towards the Alpha stage,
+work may start as a PoC during the design and as adjustments between
+the release stages.
+
+#### Releases
+- **Alpha**:
+  An initial release of the feature for experimental purposes.
+  Recommended for non-production usages, evaluation or testing.
+
+  The API is considered unstable and may change significantly.
+
+  The period in which a feature can remain in Alpha is limited,
+  assuring features are not pilling up without control.
+
+  The feature functionality is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **Beta**:
+  The first release that can be evaluated with care in production.
+  Acting as a pre-release, its main objective is to collect feedback
+  from users to assure its usefulness and readiness for graduation.
+  If there is no confidence of usage or usefulness, it may remain in
+  this stage for some time.
+  However, the period in which a feature can remain in Beta is limited,
+  assuring features are not pilling up without control.
+
+  The API is considered stable with care no not break forward compatability.
+  This implies that only additional fields may be added during this stage
+  i.e. it is not allowed to change a field path/name or remove it.
+
+  The feature functionality is controlled using a Feature-Gate (FG) during
+  runtime. It must be specified for the feature to be active.
+
+- **GA**:
+  The feature graduated to general-availability (ga) and is now part of
+  the core features.
+
+  The feature functionality is no longer controlled by a FG.
```

Comment:
It is worth noting here that there are features that need to be behind a configuration flag that needs to be turned on or turned off even in GA. This configuration flag is different from the feature life cycle flag and should be treated as such. 

Additionally this new configuration flag could be introduced at any point in the lifecycle of the feature, provided that the default behavior of the configuration flag remains the same as it was in Beta and GA releases

Reply:
I agree that FG shouldn't be used for enabling or disabling a feature in the custer. However, we do have such cases. I introduced something similar for SCSI Persistent Reservation (my bad), but there are other features like this. Could we investigate which features are misuing the feature gate in this way? At least, we would need to introduce a configuration option.

Reply:
FG's are guarding a complete feature.
Config switches (in short config) are configuring a feature.

Pre-GA, an FG is required to contain/shield/guard a feature.
In Alpha it is OFF by default
In Beta it (should) be ON by default
In GA it is gone.

For i.e. scsi persistent reservations, if this is something a cluster admin should opt-in to, then starting with alpha you want to have a cluster level configuration to enable or disable this functionality.

Reply:
There is another case that needs to be discussed here. What is the expected behavior for the following user story?

1. User has a kubevirt version installed
2. User has turned off feature foo which is in beta, using FG. Due to this the string field `vmi.spec.foo` is empty
3. User upgrades kubevirt to new version where feature foo has GA'ed. 
4. Will user automatically see the feature enabled? if so, what happens to the `vmi.spec.foo`? If it remains empty, it is not a valid configuration/input for feature foo. Alternatively feature in GA will have to handle empty field.

Does this contradicts the contract in some way? i.e. the design doc says that once feature goes from beta to GA we will not be breaking it, but the act of removing the feature gate has broken users who turned off use the feature gate in beta.

Reply:
Once Beta is released, there is a contract commitment to support upgrades to future Beta release changes and GA with backward compatibility in mind. This is explicitly mentioned in the Beta section:

>   The API is considered stable with care no not break backward compatibility
>   with previous beta releases.
>   This implies that fields may only be added during this stage,
>   not removed or renamed.

I will try to clarify further that this applies also to upgrades to GA: Just added under the GA item a sentence about it.

Reply:
Generally speaking: The design needs to cover for cases like this.
It sounds a little like `foo` would be a new mandatory feature, which should be handeled with even more care.

Reply:
>   The feature presence is controlled using a Feature-Gate (FG) during
  runtime. It must be specified for the feature to be active.
  
 Having this line in the proposal addresses my concerns, going to resolve this thread